File: occasum-core/src/main/java/pw/stamina/occasum/node/instructions/PropertyNodeInstructionList.java
package pw.stamina.occasum.node.instructions;

import pw.stamina.occasum.PropertyHandle;
import pw.stamina.occasum.properties.Property;

import java.util.Optional;

public interface PropertyNodeInstructionList extends Iterable<PropertyNodeInstruction> {

    Optional<PropertyNodeInstructionList> findNestedInstructions(String name);

    void property(PropertyHandle handle, Property property);

    void property(Property property);

    PropertyNodeInstructionList folder(PropertyHandle handle, String name);

    PropertyNodeInstructionList folder(String name);

    static PropertyNodeInstructionList newStandard() {
        return new StandardPropertyNodeInstructionList();
    }
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/node/instructions/PropertyNodeInstructionList.kt
package pw.stamina.occasum.node.instructions

import pw.stamina.occasum.PropertyHandle
import pw.stamina.occasum.properties.Property

interface PropertyNodeInstructionList : Iterable<PropertyNodeInstruction> {

    fun findNestedInstructions(name: String): PropertyNodeInstructionList?

    fun property(handle: PropertyHandle, property: Property<*>)

    fun property(property: Property<*>)

    fun folder(handle: PropertyHandle, name: String): PropertyNodeInstructionList

    fun folder(name: String): PropertyNodeInstructionList

    companion object {

        fun newStandard(): PropertyNodeInstructionList {
            return StandardPropertyNodeInstructionList()
        }
    }
}
